pilot/pkg/model/authentication.go

```diff
@@ -245,3 +250,14 @@ func constructgRPCCallCredentials(tokenFileName, headerKey string) []*core.GrpcS
 		},
 	}
 }
+
+// marshalDeterministicAny takes the protocol buffer and encodes it into google.protobuf.Any using
+// deterministic order.
+func marshalDeterministicAny(config proto.Message) *types.Any {
```

Comment:
Any chance we can test the deterministic behavior in unit test? Just in case someone break it accidentally going forward.

Reply:
And ideally this could be in some utils directory and everyone can call in Istio.

Reply:
It turns out that FileBasedMetadataConfig does not support deterministic marshal. marshalDeterministicAny() returns an err and empty value.

I use a map to stored FileBasedMetadataConfig, for the same tokenFileName, headerKey, FileBasedMetadataConfig is created and marshaled once. Following calls to findOrMarshalFileBasedMetadataConfig() will reuse this copy.